File: src/socket/src/core/ai/requests/LangflowRequest.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { createOneTimeToken } from "@/core/ai/BotOneTimeToken";
import BaseRequest, { IRequestData, IRequestExecuteParams, IRequestParams } from "@/core/ai/requests/BaseRequest";
import { LangboardCalledVariablesComponent, LangboardFile } from "@/core/ai/helpers/TweaksComponent";
import SnowflakeID from "@/core/db/SnowflakeID";
import { api } from "@/core/helpers/Api";
import Logger from "@/core/utils/Logger";
import { EHttpStatus } from "@langboard/core/enums";
import { Utils } from "@langboard/core/utils";
import formidable from "formidable";
import fs from "fs";
import { LangflowStreamResponse } from "@/core/ai/responses/LangflowResponse";
import BaseStreamResponse from "@/core/ai/responses/BaseStreamResponse";
import { EBotPlatform, EBotPlatformRunningType } from "@langboard/core/ai";

class LangflowRequest extends BaseRequest {
    protected createRequestData({ requestModel, useStream }: IRequestExecuteParams): IRequestData | null {
        const sessionId = requestModel.sessionId ?? Utils.String.Token.generate(32);
        const oneTimeToken = createOneTimeToken(new SnowflakeID(requestModel.userId));

        const queryParams = new URLSearchParams({
            stream: useStream ? "true" : "false",
        });

        let url = this.baseURL;
        switch (this.internalBot.platform_running_type) {
            case EBotPlatformRunningType.Endpoint:
                url = `${url}/${this.internalBot.value?.startsWith("/") ? this.internalBot.value.slice(1) : this.internalBot.value}`;
                break;
            case EBotPlatformRunningType.FlowJson:
                url = `${url}/api/v1/run/${this.internalBot.id}`;
                break;
        }

        url = `${url}?${queryParams.toString()}`;

        requestModel.tweaks = requestModel.tweaks ?? {};

        if (requestModel.filePath) {
            requestModel.tweaks = {
                ...requestModel.tweaks,
                ...new LangboardFile(requestModel.filePath).toTweaks(),
            };
        }

        const component = new LangboardCalledVariablesComponent(
            "chat",
            oneTimeToken,
            "user",
            { uid: new SnowflakeID(requestModel.userId).toShortCode() },
            requestModel.projectUID,
            [],
            requestModel.restData
        );

        const reqData = {
            input_value: requestModel.message,
            input_type: requestModel.inputType,
            output_type: requestModel.outputType,
            session: sessionId,
            session_id: sessionId,
            run_type: "internal_bot",
            uid: this.internalBot.uid,
            tweaks: {
                ...requestModel.tweaks,
                ...component.toTweaks(),
                ...component.toData(),
            },
        };

        return {
            url,
            oneTimeToken,
            reqData,
        };
    }

    protected createStreamResponse({ requestModel, headers, task }: Omit<IRequestParams, "useStream">): BaseStreamResponse {
        const [abortController, finish] = task ?? [undefined, undefined];

        return new LangflowStreamResponse({
            url: requestModel.url,
            headers: headers,
            body: requestModel.reqData,
            signal: abortController?.signal,
            onEnd: finish,
        });
    }

    protected convertResponse(data: { session_id: string; outputs: Record<string, any>[] }): string {
        try {
            let responseOutputs = data.outputs[0];
            while (!responseOutputs.messages) {
                responseOutputs = responseOutputs.outputs[0];
            }
            return responseOutputs.messages[0].message;
        } catch {
            return "";
        }
    }

    public async upload(file: formidable.File): Promise<string | null> {
        const filename = file.originalFilename || file.newFilename;
        if (!filename || !file.size) {
            return null;
        }

        if (![EBotPlatform.Langflow].includes(this.internalBot.platform)) {
            return null;
        }

        const headers = {
            "Content-Type": "multipart/form-data",
            ...this.getBotRequestHeaders(),
        };

        const url = `${this.baseURL}/api/v2/files`;

        const formData = new FormData();
        const blob = new Blob([fs.readFileSync(file.filepath)], { type: file.mimetype ?? undefined });
        formData.append("file", blob, filename);

        try {
            const response = await api.post(url, formData, {
                headers: {
                    ...headers,
                },
                data: formData,
            });

            if (![EHttpStatus.HTTP_200_OK, EHttpStatus.HTTP_201_CREATED].includes(response.status)) {
                throw new Error("Langflow file upload failed");
            }

            return response.data.path ?? null;
        } catch (error) {
            Logger.error(error);
            return null;
        }
    }

    public async isAvailable(): Promise<bool> {
        try {
            const healthCheck = await api.get(`${this.baseURL}/health`, {
                headers: this.getBotRequestHeaders(),
            });

            return healthCheck.status === EHttpStatus.HTTP_200_OK;
        } catch {
            return false;
        }
    }
}

export default LangflowRequest;

File: src/api/langboard/tasks/bots/utils/requests/LangflowRequest.py
from json import dumps as json_dumps
from json import loads as json_loads
from core.utils.Converter import json_default
from models import BotLog
from models.BaseBotModel import BotPlatformRunningType
from models.bases import BaseBotLogModel
from .....ai import LangboardCalledVariablesComponent
from .....core.logger import Logger
from ..BotTaskDataHelper import BotTaskDataHelper
from .BaseBotRequest import BaseBotRequest, RequestData


logger = Logger.use("BotTask")


class LangflowRequest(BaseBotRequest):
    def create_request_data(self, bot_log: tuple[BotLog, BaseBotLogModel | None]):
        project_uid = self._project.get_uid() if self._project else None
        session_id = f"{self._bot.get_uid()}-{project_uid}"
        if session_id.endswith("-"):
            session_id = session_id[:-1]

        log, scope_log = bot_log

        component = LangboardCalledVariablesComponent(
            event=self._event,
            app_api_token=self._bot.app_api_token,
            project_uid=project_uid,
            current_runner_type="bot",
            current_runner_data=BotTaskDataHelper.create_user_or_bot(self._bot),
            rest_data=json_loads(json_dumps(self._data, default=json_default)),
        )

        request_data = {
            "input_value": "",
            "input_type": "chat",
            "output_type": "chat",
            "session": session_id,
            "session_id": session_id,
            "run_type": "bot",
            "uid": self._bot.get_uid(),
            "project_uid": project_uid,
            "log_uid": log.get_uid(),
            "scope_log_table": scope_log.__tablename__ if scope_log else None,
            "tweaks": {
                **component.to_data(),
                **component.to_tweaks(),
            },
        }

        url = self._base_url
        if self._bot.platform_running_type == BotPlatformRunningType.Endpoint:
            url = f"{self._base_url}/{self._bot.value.lstrip('/')}"
        elif self._bot.platform_running_type == BotPlatformRunningType.FlowJson:
            url = f"{self._base_url}/api/v1/webhook/{self._bot.id}"

        return RequestData(
            {
                "url": url,
                "data": request_data,
            }
        )
